### PR TITLE
Refactored shard closure verification performed by ShutdownTask

### PR DESCRIPTION
**Pull request by parijatsinha**
*Issue #, if available:*
https://github.com/awslabs/amazon-kinesis-client/issues/675

*Description of changes:*
Refactored shard closure verification performed by ShutdownTask

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/ShardClosureVerificationResponse.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.Getter;
+
+import java.util.List;
+
+/**
+ * ShutDown task verifies if the shard in context is necessarily closed before checkpointing the shard at SHARD_END.
+ * The verification is done by IKinesisProxy and this class is used to wrap the verification results.
+ */
+@Getter
+public class ShardClosureVerificationResponse {
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/ShardClosureVerificationResponse.java]: Can we make ShardClosureVerificationResponse an interface which only has a method isShardClosed() and use this in IKinesisProxy. Any concrete impl like  KinesisProxy can choose to return an impl of ShardClosureVerificationResponse. This way we dont need to tie listofshards to the response, as we will be deprecating this in future.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/IKinesisProxy.java [hunk 2]
```diff
@@ -78,6 +79,18 @@ GetRecordsResult get(String shardIterator, int maxRecords)
...
 
+    /**
+     * Used to verify during ShardConsumer shutdown if the provided shardId is for a shard that has been closed.
+     * Returns the list of shards so it can be used for lease creation (instead of calling getShardList() again)
+     * along with the result of verification.
+     * Note that DynamoDBStreamsProxy does not implement this check and hence resorts to the default response.
+     * @param shardId Id of the shard that needs to be verified.
+     * @return an Object of type ShardClosureVerificationResponse.
+     */
+    default ShardClosureVerificationResponse verifyShardClosure(String shardId) {
```
[the remark this answers is not in the document]
Comment (parijatsinha) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/IKinesisProxy.java]: Addressed.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java [hunk 3]
```diff
@@ -476,6 +477,35 @@ private boolean cacheNeedsTimeUpdate() {
...
+    private boolean isShardParentOfAny(String shardId, List<Shard> shards) {
+        for(Shard shard : shards) {
+            if (isChildShardOfShard(shard, shardId)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isChildShardOfShard(Shard shard, String shardId) {
```
[the remark this answers is not in the document]
Comment (parijatsinha) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java]: Changed

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/ShardClosureVerificationResponse.java (hunk 1 above)
Comment (parijatsinha) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/ShardClosureVerificationResponse.java]: Refactored.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java [hunk 4]
```diff
@@ -107,11 +104,13 @@ public TaskResult call() {
...
              */
             if(localReason == ShutdownReason.TERMINATE) {
-                latestShards = kinesisProxy.getShardList();
+                ShardClosureVerificationResponse shardClosureVerificationResponse = kinesisProxy.verifyShardClosure(shardInfo.getShardId());
+                latestShards = shardClosureVerificationResponse.getLatestShards();
 
-                //If latestShards is null or empty, we should still shut down the ShardConsumer with Zombie state which avoid
-                // checking point with SHARD_END sequence number.
-                if(CollectionUtils.isNullOrEmpty(latestShards) || !isShardInContextParentOfAny(latestShards)) {
+                // If latestShards is null or empty, it means shard in context is not closed yet.
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: I latestShards is null or empty, we do not have enough info to take a decision. Hence we go ahead and shut downt he consumer

**Pull request merged**